phpunit/class-rest-widget-types-controller-test.php

```diff
@@ -0,0 +1,254 @@
+<?php
+/**
+ * WP_REST_Widget_Types_Controller tests.
+ *
+ * @package WordPress
+ * @subpackage REST_API
+ * @since x.x.0
+ */
+
+/**
+ * Tests for WP_REST_Widget_Types_Controller.
+ *
+ * @since x.x.0
+ *
+ * @covers WP_REST_Widget_Types_Controller
+ *
+ * @group restapi-widgets
+ * @group restapi
+ */
+class REST_Widget_Types_Controller_Test extends WP_Test_REST_Controller_Testcase {
+
+	/**
+	 * Admin user ID.
+	 *
+	 * @since x.x.0
+	 *
+	 * @var int $subscriber_id
+	 */
+	protected static $admin_id;
+
+	/**
+	 * Subscriber user ID.
+	 *
+	 * @since x.x.0
+	 *
+	 * @var int $subscriber_id
+	 */
+	protected static $subscriber_id;
+
+	/**
+	 * Create fake data before our tests run.
+	 *
+	 * @since x.x.0
+	 *
+	 * @param WP_UnitTest_Factory $factory Helper that lets us create fake data.
+	 */
+	public static function wpSetUpBeforeClass( $factory ) {
+		self::$admin_id      = $factory->user->create(
+			array(
+				'role' => 'administrator',
+			)
+		);
+		self::$subscriber_id = $factory->user->create(
+			array(
+				'role' => 'subscriber',
+			)
+		);
+	}
+
+	public static function wpTearDownAfterClass() {
+		self::delete_user( self::$admin_id );
+		self::delete_user( self::$subscriber_id );
+	}
+
+	/**
+	 *
+	 */
+	public function test_register_routes() {
+		$routes = rest_get_server()->get_routes();
+		$this->assertArrayHasKey( '/__experimental/widget-types', $routes );
+		$this->assertCount( 1, $routes['/__experimental/widget-types'] );
+		$this->assertArrayHasKey( '/__experimental/widget-types/(?P<name>[a-zA-Z0-9_-]+)', $routes );
+		$this->assertCount( 2, $routes['/__experimental/widget-types/(?P<name>[a-zA-Z0-9_-]+)'] );
+	}
+
+	/**
+	 *
+	 */
+	public function test_context_param() {
+		// Collection.
+		$request  = new WP_REST_Request( 'OPTIONS', '/__experimental/widget-types' );
+		$response = rest_get_server()->dispatch( $request );
+		$data     = $response->get_data();
+		$this->assertSame( 'view', $data['endpoints'][0]['args']['context']['default'] );
+		$this->assertSame( array( 'view', 'embed', 'edit' ), $data['endpoints'][0]['args']['context']['enum'] );
+		// Single.
+		$request  = new WP_REST_Request( 'OPTIONS', '/__experimental/widget-types/calendar' );
+		$response = rest_get_server()->dispatch( $request );
+		$data     = $response->get_data();
+		$this->assertSame( 'edit', $data['endpoints'][0]['args']['context']['default'] );
+		$this->assertSame( array( 'view', 'embed', 'edit' ), $data['endpoints'][0]['args']['context']['enum'] );
+	}
+
+	/**
+	 * @expectedIncorrectUsage rest_sanitize_value_from_schema
```

Comment:
What is causing us to issue `_doing_it_wrong` warnings?

Reply:
I didn't work that out, was going to look at this tommorrow.

Reply:
Worked it out.